notebooks/_06-Architecture.md

```diff
@@ -0,0 +1,429 @@
+Code for input validation can be built in separate functions and be used part of analysis functions as seen needed.
+
+### 13.7 How will we use (or not use) type hinting to help prevent errors?
+Docstrings will suggest to the users what each of the function takes in as types and what it returns.
+
+## 14. Description of testing plan:
+### 14.1 How will we test that all of the parts of the architecture are working correctly?
+- **How will we create tests for each part of the architecture that are:**
+ - **clear and unambiguous**
+ - **capable of dealing with stochastic functions / objects**
+ - **minimizing reliance / use of stochasticity**
+ - **fast to evaluate**
+
+This will be filled in as we learn more about tests. The first thing that comes to mind is running the same tests
+many times and then verify that the output statistics match one's expectations. In order for tests to be fast, each
+test will need to focus on one functionality. We would need to use descriptive names to understand what the tests are
+doing. Each test should also be independent from other tests.
+
+### 14.2 What are the various types of tests that are needed?
+Unit tests/Feature tests, integration tests.
+
+### 14.3 Are there any parts of the architecture that we do not know how to test? How will these knowledge gaps be resolved before the coding begins?
+Given our preliminary knowledge on testing, there will be elements that we won't know how to test. 
+Reading through available documentation and tutorials for pytest or other packages will help resolve any knowledge gaps 
+in testing any of the parts of the architecture.
+
+### 14.4 How will discovered bugs be turned into test cases?
+The guide provided by timothy on how to debug code will be followed to move from finding the bug to testing it.
+
+### 14.5 How will the software be reviewed in general (beyond automated tests) to reduce the probability of programming bugs into the code / increase the chance of producing correctly functioning software?
+cross-review between team members will help in detecting any bugs.
+
+## 15. Description of project-sharing procedures:
+### 15.1 Is there a plan in place for producing documentation for the software?
+- **Documentation for operating the software**
+- **Documentation of software limits**
+- **Usage examples and tutorials**
+- **Key project learnings / retrospective notes**
+
+Documentation, in the form of Jupyter Notebook accompanied by a Markdown file (README.md) detailing the procedure for installing the environment and the content of folders and notebooks. Any limitations to the project will be noted in the jupyter notebooks developed to illustrate the analysis effort. The README.md markdown file will also will point to documentation pages for any packages used in the project.
```

Comment:
In addition to what you've written, we should DEFINITELY include docstrings for all public (i.e. non-underscored) functions used in the project.

A user should be able to call help on any function or class method we've written and learn all they need to know to use the function as intended.